### PR TITLE
improve version check speed for already downloaded versions

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
As part of making it so I can locally test [AB#8548](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8548)

I haven't tested this for remote builds, but this sped up my local runs of loader back-compat downloads significantly from minutes to a few seconds. 

Even though the remote builds will need to download all the builds, this speeds up the process of reading whether or not a build was installed. My guess is that there likely will be pipeline speed reductions.

Locally, running back compat full tests will be significantly faster as indicated above because we will not need to read the file for every single version that we have installed, instead we will read the file once for all installed versions.

### CODE REVIEW

**Review by jatgarg — APPROVED** [review]

**Review by Abe27342 — CHANGES_REQUESTED** [review]
this doesn't look safe to me. have you tested that it functions properly when various bits of the install tree and/or installed.json changes? I'm worried that cached result doesn't remain accurate.

I do agree that some optimization along these lines can be performed and is worth doing, but I'm not sure this one works.

**Comment by tyler-cai-microsoft** [comment]
> this doesn't look safe to me. have you tested that it functions properly when various bits of the install tree and/or installed.json changes? I'm worried that cached result doesn't remain accurate.
> 
> I do agree that some optimization along these lines can be performed and is worth doing, but I'm not sure this one works.

I did try deleting my local files completely, I'm going to check to see if the remote works.

**Comment by Abe27342** [comment]
> > this doesn't look safe to me. have you tested that it functions properly when various bits of the install tree and/or installed.json changes? I'm worried that cached result doesn't remain accurate.
> > I do agree that some optimization along these lines can be performed and is worth doing, but I'm not sure this one works.
> 
> I did try deleting my local files completely, I'm going to check to see if the remote works. Regardless, I need this locally, because I can't make the necessary changes unless I do this.

what do you mean local/remote in this context?

also, I'm interested also in robustness around partial deletions / incremental breaks (e.g. you have the right versions installed today, then 2 weeks from now you pull and we test a different compat matrix because we've upgraded to 2.1 or whatever--does everything work correctly?). That kind of issue is pretty rough to debug if you don't have context on what this code is doing.

**Comment by tyler-cai-microsoft** [comment]
> > > this doesn't look safe to me. have you tested that it functions properly when various bits of the install tree and/or installed.json changes? I'm worried that cached result doesn't remain accurate.
> > > I do agree that some optimization along these lines can be performed and is worth doing, but I'm not sure this one works.
> > 
> > 
> > I did try deleting my local files completely, I'm going to check to see if the remote works. Regardless, I need this locally, because I can't make the necessary changes unless I do this.
> 
> what do you mean local/remote in this context?
> 
> also, I'm interested also in robustness around partial deletions / incremental breaks (e.g. you have the right versions installed today, then 2 weeks from now you pull and we test a different compat matrix because we've upgraded to 2.1 or whatever--does everything work correctly?). That kind of issue is pretty rough to debug if you don't have context on what this code is doing.

All your points are valid, and I do have concerns about them. I don't mind moving this back to draft form until it's thoroughly tested. I got it to work on my machine locally, but I think it's because I kind of understand what's going on here.

**Comment by Abe27342** [comment]
The places where we modify the set of installed packages is quite small; `addInstalled` and `removeInstalled` in this file (not exported). I think you could probably make a small tweak to the spirit of this PR so that we only recompute `readInstalledJson` on modifications to it (but still do so more than once), and then this seems ok.

**Comment by tyler-cai-microsoft** [comment]
> The places where we modify the set of installed packages is quite small; `addInstalled` and `removeInstalled` in this file (not exported). I think you could probably make a small tweak to the spirit of this PR so that we only recompute `readInstalledJson` on modifications to it (but still do so more than once), and then this seems ok.

What I really want to do is to turn this into an object that manages its own state, but that's a lot of change.

File: packages/test/test-version-utils/src/versionUtils.ts [hunk 1]
```diff
@@ -40,6 +40,12 @@ interface InstalledJson {
 	installed: string[];
 }
 
+let cachedInstalledJson: InstalledJson | undefined;
```
Comment (tyler-cai-microsoft) [inline packages/test/test-version-utils/src/versionUtils.ts:43]: What do you think of this whole caching thing? I was thinking of turning this whole thing into a whole object, but that might be a refactoring for later.

Comment (tyler-cai-microsoft) [reply packages/test/test-version-utils/src/versionUtils.ts:43]: @Abe27342

Comment (Abe27342) [reply packages/test/test-version-utils/src/versionUtils.ts:43]: seems ok to me. We do kind of now have split-brain here around how we manage the state of currently installed packages, but i don't think the before state is particularly elegant and it's all implementation details, so easy enough to change when we have time to improve further.

**Review by Abe27342 — APPROVED** [review]

**Comment by tyler-cai-microsoft** [comment]
From manual testing, I deleted `2.0.0-internal.3.0.11` and ran the loader compat test with full compat enabled, and it installed the version and ran the test.

**Pull request merged**